### PR TITLE
Reconciler unit tests

### LINKED ISSUE #1241: Templating controller missing tests
### What problem are you facing?

The generic templating controller used in template stacks does not have as much test coverage as original crossplane repositories.

https://github.com/crossplaneio/resourcepacks (will be renamed to `templating-controller` once https://github.com/crossplaneio/resourcepacks/pull/2 is merged)

### How could Crossplane help solve your problem?

Implement unit tests in that repository.

### PR DESCRIPTION
**Pull request by muvaf**
Introduces unit tests for the main reconciler and all of the patchers and meta functions.

Fixes https://github.com/crossplaneio/crossplane/issues/1241

### CODE REVIEW

File: pkg/resource/meta_test.go
Comment (displague) [inline pkg/resource/meta_test.go:97]: I thought we might be able to avoid adding `SetConditions` tests here if we made use of [v1alpha1.SetConditions](https://github.com/crossplaneio/crossplane-runtime/blob/master/apis/core/v1alpha1/condition.go#L150) in [this location](https://github.com/crossplaneio/templating-controller/pull/6/files#diff-9d507b08295a03cb40e6423b7b9c729fL74).

But even with that, SetCondition in this project has additional responsibilities. :+1:

**Review by displague — APPROVED** [review]
Coverage increased from 23.5% to 78.4% in `controllers`, and 0% to 72.6% in `resource`. :clinking_glasses: 

For your efforts, we are less savage.

**Review by suskin — COMMENTED** [review]
Awesome! :squid:

The tests that are here look pretty good to me. My questions are all about tests that aren't here yet : )

* Looks like we have good coverage of success cases and client call failures. What are your thoughts on testing the handling of malformed stack definitions or stacks? I don't think I saw any of those
* What are your thoughts on tests for the code in the kustomize and helm3 packages?

**Review by suskin — APPROVED** [review]
Approving assuming the conflicts are resolved. Regarding my question about other tests, I figure we can always add more tests in another PR, so we don't need to block this PR on adding more tests.

**Comment by muvaf** [comment]
> What are your thoughts on testing the handling of malformed stack definitions or stacks? I don't think I saw any of those

We're using the fields of `StackDefinition` as inputs to engine options. I think we'll have tests for engine options but I'm not sure how we'd test for malformed stack definitions (or whether we need to). Do you have some test cases in mind you can share?

> What are your thoughts on tests for the code in the kustomize and helm3 packages?

I wanted to save them for another PR. When you look at them, it's actually a few kustomize patchers that we can write unit tests for. For actual templating behavior(`Run` function), I need to think more about how we can mock kustomize/helm3, otherwise we might end up testing kustomize/helm3 themselves.

**Comment by suskin** [comment]
> Do you have some test cases in mind you can share?

@muvaf Yeah definitely. Here are some ideas:

* The source referenced in the stack definition has no resources in it
* The source referenced in the stack definition has unparseable yaml files
* The engine type in the stack definition is not a supported type
* There is no engine configuration at all
* The `path` for the source doesn't exist
* The crd in the stack definition doesn't exist

**Pull request merged**
